### DISCUSSION TITLE
prototype Npc asc armor

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#351] [Show and tell]
Хочу проработать прототипы НПС для фракций Менты / Наёмы / Палы и т.д.
Пришлось сделать таблицы что бы соблюсти баланс

https://docs.google.com/spreadsheets/d/1tspBhvi9wPY2TOAkHUQTQgijPHVgt1KjjaMSHiy5WDw

Сами прототипы пока добавил в один файл
https://github.com/Kor-Angar/G2_NotR_Kor_Angar/blob/master/PrjGOTHIC/Story/NPC_Scripts/Npc_Default.d

**Comment by Kor-Angar** [comment]
А я всё с тем же вопросом :-) может есть какие то замечания? :-)
```cpp
prototype NPC_SLD_lvl_40_2h(C_Npc)
{	
	//-------- primary data --------
	name[0] = 						NAME_Soeldner;
	npctype =						NPCTYPE_AMBIENT;
	guild =							GIL_SLD;
	level =							40;
	flags =							0;			
	//-------- abilities --------
	/*
	//	взято из B_SetAttributesToChapter(4)
		slf.level = 40;
		slf.attribute[ATR_STRENGTH] = 150;
		slf.aivar[REAL_STRENGTH] = 150;
		slf.attribute[ATR_DEXTERITY] = 150;
		slf.aivar[REAL_DEXTERITY] = 150;
		slf.attribute[ATR_MANA_MAX] = 1000;
		slf.attribute[ATR_MANA] = 1000;
		slf.aivar[REAL_MANA_MAX] = 1000;
		slf.attribute[ATR_HITPOINTS_MAX] = 520;
		slf.attribute[ATR_HITPOINTS] = 520;
	*/
	attribute[ATR_STRENGTH]  =		150;
	aivar[REAL_STRENGTH]	 = 		150;
	attribute[ATR_DEXTERITY] =		150; // а надо ли столько?
	aivar[REAL_DEXTERITY]	 = 		150;
	attribute[ATR_MANA_MAX] =		10;	 // срезал до 10
	attribute[ATR_MANA] =			10;	 // срезал до 10
	attribute[ATR_HITPOINTS_MAX] =	520;
	attribute[ATR_HITPOINTS] =		520;
	// ------ Grund-Schadensresistenzen: Keine ------
	protection	[PROT_EDGE]				= 0;
	protection	[PROT_BLUNT]			= 0;
	protection	[PROT_POINT]			= 0;
	protection	[PROT_FIRE]				= 0;
	protection	[PROT_MAGIC]			= 0;
	//-------- visuals --------
	// прототипу делать вредно и опасно!!!
	//---------------------------	
	fight_tactic	=	FAI_HUMAN_STRONG;	// MASTER / STRONG / COWARD
	
	
	//-------- Talente -------- 
	// в оригинале было 	B_SetFightSkills(self,50);
	HitChance[NPC_TALENT_1H] 			 = 30;		 //¬ладение ќдноручником 
	HitChance[NPC_TALENT_2H] 	    	 = 50;		 //¬ладение ƒвуручником 
	HitChance[NPC_TALENT_BOW]			 = 50;		 //”мение обращатьс¤ с Ћуком 
	HitChance[NPC_TALENT_CROSSBOW]		 = 10;		 //”мение обращатьс¤ с јрбалетом
	
	Npc_SetTalentSkill(self,NPC_TALENT_1H,1);		 // 0-обучен, 1 - боец, 2 - мастер
	Npc_SetTalentSkill(self,NPC_TALENT_2H,1);		 // 0-обучен, 1 - боец, 2 - мастер
	Npc_SetTalentSkill(self,NPC_TALENT_BOW,1);		 // 0-обучен, 1 - боец, 2 - мастер
	Npc_SetTalentSkill(self,NPC_TALENT_CROSSBOW,0);	 // 0-обучен, 1 - боец, 2 - мастер
		
		
	// ------ Schdenstyp im Faustkampf ------
	damagetype 							= DAM_BLUNT;

	// ------ senses ------
	senses			=	SENSE_HEAR | SENSE_SEE ;
	senses_range	=	PERC_DIST_ACTIVE_MAX;	//in AI_Constants.d
		
	// ----- aivar Defaultwerte  ------
	aivar[AIV_MM_FollowTime] = NPC_TIME_FOLLOW;
	//aivar[AIV_FightDistCancel] = FIGHT_DIST_CANCEL;
	//-------- inventory --------
	EquipItem(self,ItMw_2h_Sld_Axe);
	B_CreateAmbientInv(self);
};


func void Set_Relaxed_Visuals()
{	
	Mdl_SetVisual		(self,	"HUMANS.MDS");
	Mdl_ApplyOverlayMds	(self,"Humans_Relaxed.mds");
//	Mdl_SetModelFatness(self,0);
//	B_Scale(self);
};

```
Инстанция получались компактная :-)

```cpp
instance SLD_826_Soeldner(NPC_SLD_lvl_40_2h)
{
	// ------ NSC ------
	id = 826;
	voice = 6;
	
	//-------- visuals --------
	B_SetNpcVisual(self,MALE,"Hum_Head_Bald",Face_N_Bullit,BodyTex_N,itar_sld_M);
	Set_Relaxed_Visuals();
	Mdl_SetModelFatness(self,1);

	// ------ TA anmelden ------		
	daily_routine = Rtn_Start_826;
};


func void Rtn_Start_826()
{
	TA_Smalltalk(8,0,22,0,"NW_BIGFARM_HOUSE_08");
	TA_Sleep(22,0,8,0,"NW_BIGFARM_HOUSE_SLD_SLEEP");
};

```

**Comment by dosinabox** [comment]
```
lvl_40_2h
```
Я бы уровни и владения выносила из прототипа в инстанции. Слишком индивидуальные вещи. На каждый уровень по прототипу, получается?

**Comment by Kor-Angar** [comment]
> На каждый уровень по прототипу, получается?

Скорее по два-три прототипа на фракцию. Сейчас  инстанции NPC формируется  через B_SetAttributesToChapter(ХХ) и B_SetFightSkills(self,YY). и это совсем неправильно.
ввиду того что в начале года я погружался в код Готики 1, то после него инстанции НПС в Готике 2 смотрится совсем-совсем шаблонный. в той же Готики 1 тамплиеры имели условно три шаблона и индивидуальные моменты у именных НПС. 
в Готике 2 НВ это свелось к условно одному (!) шаблону  для наёмников: B_SetAttributesToChapter(4) и B_SetFightSkills(self,50)
Для наглядности я [вбил наёмников в табличку](https://docs.google.com/spreadsheets/d/1tspBhvi9wPY2TOAkHUQTQgijPHVgt1KjjaMSHiy5WDw/edit?usp=sharing)
![image](https://user-images.githubusercontent.com/49514764/93417527-0da46980-f8b1-11ea-9f07-b028e774abe1.png)

тут видно сразу что у неименных НПС всё одинаково: только доспехи отличаются. хотя вроде как доспех предполагает что НПС более сильный. в принципе на наёмников хватит  и одного прототипа, если не делать специализацию на 1р или 2р (сейчас у всех НПС 1р=2р=лук=арбалет что мягко говоря странно).


> Я бы уровни и владения выносила из прототипа в инстанции

а зачем их именно выносить? пусть будут в прототипе, если НПС такой весь "индивидуальный", как например Ярвис, то ничего не мешает ему в  инстанции  скорректировать и уровень и владения арбалетом и шестой круг магии и 1000 маны. ведь прототип на то и прототип? :-)

**Comment by Kor-Angar** [comment]
попробовал как то расставить НПС в табличке по "качеству"

![image](https://user-images.githubusercontent.com/49514764/93419110-95d83e00-f8b4-11ea-856a-7f791626221c.png)

а у тебя в правках Бастер стал супер мечником? :-)

**Comment by dosinabox** [comment]
> а у тебя в правках Бастер стал супер мечником? :-)

Произошло повышение 1р с 30 до 60, потому что он учит до 60.

**Comment by Kor-Angar** [comment]
> Произошло повышение 1р с 30 до 60, потому что он учит до 60.

так проще ему порог обучения было до 30 опустить. Основной учитель Корд, а у Бастера тренируется минимальному уровню, что бы Корд дал голос и согласился учить ГГ.

Бастер и Фестер были самыми слабыми наемниками в игре, что бы низкоуровневый герой мог их побить без проблем. Обучение у Бастера по моему(могу конечно ошибаться) доступно только если его побить. 

Сейчас получается что ты этой правкой фактически усложнила вступление в наёмники.

**Comment by dosinabox** [comment]
> Сейчас получается что ты этой правкой фактически усложнила вступление в наёмники.

В таком случае, в следующей версии Бастер будет единственным, кто учит большему (60), чем знает сам (30).

**Comment by Kor-Angar** [comment]
он будет не единственным :-) такая же ситуация и у Корда и у Рода и у Вульфгара и у Волка и у большинство учителей. Это связано как раз  тем что Пираньи расслабились и стали инстанции формировать через B_SetAttributesToChapter(Х) и B_SetFightSkills(self,YY).

что бы решить момент с учителями, нужно решить более глобальную задачу - делать нормальные прототипы для Ментов / Наёмов / Палов / Бандитов. Про как раз это тема :-)

Тогда у Наёмников не будет 6 кругов магии, 1000 мана 50 лука/арбалета/1р/2р :-) а будет как то адекватно осознанно. При этом естественно надо делать это так что бы не нарушился баланс (кол-во опыта, сложность и т.д.), что совсем и совсем не просто, так как требует не только расчётов, но и многочисленных тестов разными игроками.
![image](https://user-images.githubusercontent.com/49514764/93522013-86e79f00-f939-11ea-9c92-7b2b9ae1d042.png)

**Comment by Kor-Angar** [comment]
финальные версии такие получились. вроде работают. но возможно чего то в них нехватает :-)

```
prototype NPC_SLD_lvl_31_1h(C_Npc)
{	
	//-------- primary data --------
	name[0] = 						NAME_Soeldner;
	npctype =						NPCTYPE_AMBIENT;
	guild =							GIL_SLD;
	level =							31;
	flags =							0;			
	//-------- abilities --------
	attribute[ATR_STRENGTH]  =		125;
	aivar[REAL_STRENGTH]	 = 		125;
	attribute[ATR_DEXTERITY] =		125;
	aivar[REAL_DEXTERITY]	 = 		125;
	attribute[ATR_MANA_MAX] =		10;	 // срезал до 10
	attribute[ATR_MANA] =			10;	 // срезал до 10
	attribute[ATR_HITPOINTS_MAX] =	400;
	attribute[ATR_HITPOINTS] =		400;
	// ------ Grund-Schadensresistenzen: Keine ------
	protection	[PROT_EDGE]				= 0;
	protection	[PROT_BLUNT]			= 0;
	protection	[PROT_POINT]			= 0;
	protection	[PROT_FIRE]				= 0;
	protection	[PROT_MAGIC]			= 0;
	//-------- visuals --------
	// прототипу делать вредно и опасно!!!
	//---------------------------	
	fight_tactic	=	FAI_HUMAN_STRONG;	// MASTER / STRONG / Normal /COWARD
	
	//-------- Talente -------- 
	B_AddFightSkill		 (self,	NPC_TALENT_1H,60);
	B_AddFightSkill		 (self, NPC_TALENT_2H,30);
	B_AddFightSkill		 (self, NPC_TALENT_BOW,60);
	B_AddFightSkill		 (self, NPC_TALENT_CROSSBOW,30);
		
	// ------ Schdenstyp im Faustkampf ------
	damagetype 							= DAM_BLUNT;

	// ------ senses ------
	senses			=	SENSE_HEAR | SENSE_SEE ;
	senses_range	=	PERC_DIST_ACTIVE_MAX;	//in AI_Constants.d
		
	// ----- aivar Defaultwerte  ------
	aivar[AIV_MM_FollowTime] = NPC_TIME_FOLLOW;
	aivar[AIV_FightDistCancel] = FIGHT_DIST_CANCEL;
	bodyStateInterruptableOverride = FALSE;

	//-------- inventory --------
	EquipItem(self,ItMw_Sld_SteinBrecher);	// одноручная камнедробилка
	EquipItem(self,ItRw_Sld_Bow);			// лук
	CreateInvItems (self, ItRw_Arrow, 50); // стрелы	
//	EquipItem(self,ItMw_1h_Sld_Axe);
//	EquipItem(self,ItMw_1h_Sld_Sword);
	B_CreateAmbientInv(self);
};


prototype NPC_SLD_lvl_31_2h(C_Npc)
{	
	//-------- primary data --------
	name[0] = 						NAME_Soeldner;
	npctype =						NPCTYPE_AMBIENT;
	guild =							GIL_SLD;
	level =							31;
	flags =							0;			
	//-------- abilities --------
	attribute[ATR_STRENGTH]  =		125;
	aivar[REAL_STRENGTH]	 = 		125;
	attribute[ATR_DEXTERITY] =		125; 
	aivar[REAL_DEXTERITY]	 = 		125;
	attribute[ATR_MANA_MAX] =		10;	 // срезал до 10
	attribute[ATR_MANA] =			10;	 // срезал до 10
	attribute[ATR_HITPOINTS_MAX] =	400;
	attribute[ATR_HITPOINTS] =		400;
	// ------ Grund-Schadensresistenzen: Keine ------
	protection	[PROT_EDGE]				= 0;
	protection	[PROT_BLUNT]			= 0;
	protection	[PROT_POINT]			= 0;
	protection	[PROT_FIRE]				= 0;
	protection	[PROT_MAGIC]			= 0;
	//-------- visuals --------
	// прототипу делать вредно и опасно!!!
	//---------------------------	
	fight_tactic	=	FAI_HUMAN_STRONG;	// MASTER / STRONG / COWARD
		
	//-------- Talente -------- 
	B_AddFightSkill		 (self,	NPC_TALENT_1H,30);
	B_AddFightSkill		 (self, NPC_TALENT_2H,60);
	B_AddFightSkill		 (self, NPC_TALENT_BOW,10);
	B_AddFightSkill		 (self, NPC_TALENT_CROSSBOW,30);			
		
	// ------ Schdenstyp im Faustkampf ------
	damagetype 							= DAM_BLUNT;

	// ------ senses ------
	senses			=	SENSE_HEAR | SENSE_SEE ;
	senses_range	=	PERC_DIST_ACTIVE_MAX;	//in AI_Constants.d
		
	// ----- aivar Defaultwerte  ------
	aivar[AIV_MM_FollowTime] = NPC_TIME_FOLLOW;
	aivar[AIV_FightDistCancel] = FIGHT_DIST_CANCEL;
	bodyStateInterruptableOverride = FALSE;

	//-------- inventory --------	
//	EquipItem(self,ItMw_2h_Sld_Axe);
	EquipItem(self,ItMw_Sld_Streitaxt); // Двуручный Топор
	
	B_CreateAmbientInv(self);
};
```

Инстанция безымянного Наёмника
```
instance SLD_817_Soeldner(NPC_SLD_lvl_31_1h)
{
	id = 817;
	voice = 6;
	
	B_SetNpcVisual(self,MALE,"Hum_Head_Bald",Face_L_Tough01,BodyTex_L,itar_sld_M);
	Mdl_SetModelFatness(self,0);
	Set_Relaxed_Visuals();
	
	daily_routine = Rtn_Start_817;
};
```
Инстанция уникального Наёмника. По ним пожалуйста сделай замечания.
```
instance SLD_811_Wolf(NPC_SLD_lvl_31_1h)
{
	name[0] = "Вольф";
	id = 811;
	voice = 8;
	level =							40;

	npcType = npctype_main;
	
	//-------- abilities --------
	attribute[ATR_STRENGTH]  =		150;
	aivar[REAL_STRENGTH]	 = 		150;
	attribute[ATR_DEXTERITY] =		150; // а надо ли столько?
	aivar[REAL_DEXTERITY]	 = 		150;
	attribute[ATR_HITPOINTS_MAX] =	520;
	attribute[ATR_HITPOINTS] =		520;
	
	B_AddFightSkill		 (self,	NPC_TALENT_1H,5);
	B_AddFightSkill		 (self, NPC_TALENT_2H,5);
	B_AddFightSkill		 (self, NPC_TALENT_BOW,30);
	B_AddFightSkill		 (self, NPC_TALENT_CROSSBOW,30);

	B_SetNpcVisual(self,MALE,"Hum_Head_FatBald",Face_N_Wolf,BodyTex_N,ItAr_Sld_m);
	Mdl_SetModelFatness(self,0);
	Set_Relaxed_Visuals();

	Npc_RemoveInvItem(self,ItMw_Sld_SteinBrecher);
	Npc_RemoveInvItem(self,ItMw_Sld_Streitaxt);
	EquipItem(self,ItMw_Nagelkeule2);

	daily_routine = Rtn_Start_811;
};
```
Тут решил сначала сбросить все навыки на 0, а потом прописать, что бы было явно видно какие параметры. Хотя можно было использовать B_AddFightSkill с отрицательным значением.
```
instance SLD_802_Buster(NPC_SLD_lvl_31_1h)
{
	name[0] = "Ѕастер";
	id = 802;
	voice = 13;
	npcType = npctype_main;
	
	//-------- abilities --------
	level =							18;
	attribute[ATR_STRENGTH]  =		100;
	aivar[REAL_STRENGTH]	 = 		100;
	attribute[ATR_DEXTERITY] =		100; // а надо ли столько?
	aivar[REAL_DEXTERITY]	 = 		100;
	attribute[ATR_HITPOINTS_MAX] =	280;
	attribute[ATR_HITPOINTS] =		280;
	
	HitChance [NPC_TALENT_1H] = 0; //¬ладение ќдноручником 
	HitChance [NPC_TALENT_2H] = 0; //¬ладение ?вуручником 
	HitChance [NPC_TALENT_BOW] = 0; //”мение обращатьс¤ с Ћуком 
	HitChance [NPC_TALENT_CROSSBOW] = 0; //”мение обращатьс¤ с јрбалетом
//	B_AddFightSkill		 (self,	NPC_TALENT_1H,1);
//	B_AddFightSkill		 (self, NPC_TALENT_2H,1);
//	B_AddFightSkill		 (self, NPC_TALENT_BOW,1);
//	B_AddFightSkill		 (self, NPC_TALENT_CROSSBOW,1);
	
	B_AddFightSkill		 (self,	NPC_TALENT_1H,30);
	B_AddFightSkill		 (self, NPC_TALENT_2H,10);
	B_AddFightSkill		 (self, NPC_TALENT_BOW,30);
	B_AddFightSkill		 (self, NPC_TALENT_CROSSBOW,10);
	
	fight_tactic = FAI_HUMAN_NORMAL;
	
	Npc_RemoveInvItem(self,ItMw_Sld_SteinBrecher);
	Npc_RemoveInvItem(self,ItMw_Sld_Streitaxt);
	EquipItem(self,ItMw_1h_Sld_Axe);
	
	B_SetNpcVisual(self,MALE,"Hum_Head_FatBald",Face_N_Drax,BodyTex_N,itar_sld_M);
	Mdl_SetModelFatness(self,0);
	Set_Relaxed_Visuals();
	
	
/*
	HitChance [NPC_TALENT_1H] = 60; //¬ладение ќдноручником 
	HitChance [NPC_TALENT_2H] = 30; //¬ладение ƒвуручником 
	HitChance [NPC_TALENT_BOW] = 30; //”мение обращатьс¤ с Ћуком 
	HitChance [NPC_TALENT_CROSSBOW] = 10; //”мение обращатьс¤ с јрбалетом
	
	Npc_SetTalentSkill(self,NPC_TALENT_1H,1); // 0-обучен, 1 - боец, 2 - мастер
	Npc_SetTalentSkill(self,NPC_TALENT_2H,0); // 0-обучен, 1 - боец, 2 - мастер
	Npc_SetTalentSkill(self,NPC_TALENT_BOW,1); // 0-обучен, 1 - боец, 2 - мастер
	Npc_SetTalentSkill(self,NPC_TALENT_CROSSBOW,0); // 0-обучен, 1 - боец, 2 - мастер
*/	
	daily_routine = Rtn_Start_802;
};


func void Rtn_Start_802()
{
	TA_Smalltalk(7,0,21,0,"NW_BIGFARM_HUT_01");
	TA_Sit_Bench(21,0,7,0,"NW_BIGFARM_HUT_BENCH_02");
};

```

**Comment by Kor-Angar** [comment]
в прототипы так же удачно можно вставить и случайное оружие :-) точнее я оружие добавил в B_CreateAmbientInv который продублировал для удобства в прототипах. это позволило сохранить ущерб (баланс) и в тоже время добавить визуальное разнообразие. у кого то меч, у кого то топор у кого то молоток :-)

**Reply by Kor-Angar** [reply]
> точнее я оружие добавил в B_CreateAmbientInv который продублировал для удобства в прототипах

в 2022 году с учётом опциональности модов это идея плохая.

**Comment by Kor-Angar** [comment]
в табличку добавил Паладинов
https://docs.google.com/spreadsheets/d/1tspBhvi9wPY2TOAkHUQTQgijPHVgt1KjjaMSHiy5WDw
вбил исходные данные. сразу бросилось в глаза пару паладинов (219, 267, 260) у которых низкие показатели. может это их забыли поправить и у них параметры остались из Готики 2 классик? как можно это проверить? :-)

так же странно что все паладины 50 уровня вооружены только одноручными мечами :-)

**Comment by Kor-Angar** [comment]
модельки

[ANIMS.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/8820965/ANIMS.zip)

**Reply by Kor-Angar** [reply]
@dosinabox вот asc

**Comment by Kor-Angar** [comment]
инстанции

[Items.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/8820976/Items.zip)

**Reply by dosinabox** [reply]
```cpp
name =	"Лёгкие Доспехи Cтражника";
```
Разве в стандартных шрифтах Г2 есть буква "ё"?

**Reply by Kor-Angar** [reply]
наверное нет.  я еще не редактировал в этом году инстанции :-)

**Reply by Kor-Angar** [reply]
эти инстанции скинул что бы ещё раз подчеркнуть что  в asc прописаны текстуры под `visual_skin`  и что это  важно :-)

**Comment by Kor-Angar** [comment]
по моему я стал еще ближе к финалу и в ближайшее пару недель выкачу в паблик :-)
https://steamcommunity.com/sharedfiles/filedetails/?id=2803160788

**Comment by Kor-Angar** [comment]
Миненталь:

- [ ] рудокопы
- [ ] паладины
- [ ] ополченцы
- [ ] онд 
- [ ] вип

```D
func void  B_Minental_NPC_All ()
{ 

	NPC_60_lvl_2h		(VLK_4148_Gestath);	
	B_EquipMeleeWeapon	(VLK_4148_Gestath,ItMw_Zweihaender3);


	B_Minental_NPC_BDT ();	// ok
	B_Minental_NPC_STRF ();	//
	B_Minental_NPC_Pal ();	// ok
	// B_Minental_NPC_Mil ();
	
	if	(Kapitel <= 2)
	{
		NPC_60_lvl_2h		(PC_Fighter_OW);	
		B_EquipArmor		(PC_Fighter_OW,ITAR_SLD_H4);
	
		NPC_60_lvl_1h		(PC_ThiefOW);
		B_EquipMeleeWeapon	(PC_ThiefOW, ItMw_Schwert3);
	}	
	else if	(Kapitel >= 4)
	{
		B_Minental_NPC_Kapitel_4 ();
	};
};
```

**Comment by Kor-Angar** [comment]
```D
func void B_EquipArmor(var C_Npc slf,var int NewArmor)
{
	if	!C_NpcCanGetItem(slf,NewArmor)
	{
		return;
	};

	if(Npc_HasEquippedArmor(slf)) //чтобы OldArmor не был null
	{		
		var C_ITEM	OldArmor;			OldArmor 			= Npc_GetEquippedArmor	(slf);
		var int		OldArmorInstance;	OldArmorInstance	= Hlp_GetInstanceID		(OldArmor);
		Npc_RemoveInvItem	(slf,OldArmorInstance);
		//или Npc_RemoveInvItem(slf,OldArmor)?
	};
	CreateInvItem		(slf,NewArmor);
	AI_EquipBestArmor	(slf);
	
	slf.aivar[AIV_MM_REAL_ID] = 222;
};
```

**Reply by Kor-Angar** [reply]
```D
func int C_NpcCanGetItem(var C_Npc slf,var int Item)
{
	if(!Hlp_IsValidNpc(slf))
	{
		return FALSE;	//npc не существует
	};
	if(Npc_HasItems(slf,Item))
	{
		return FALSE;	//у npc уже есть предлагаемый Item
	};
	return TRUE;
};
```

**Comment by Kor-Angar** [comment]
```D
// ***********************************************
// B_RefreshAtInsert
// -----------------
// Funktion wird von der Engine aufgerufen
// und zwar fьr jedem NSC (auch Monster), 
// die in die "Darstellungsglocke" kommen
// self ist hier immr der jeweils abgehandelte NSC
// ***********************************************

func void B_RefreshAtInsert()
{
    // EXIT IF...
    
    // ------ Tote = Exit ------
    if (self.attribute[ATR_HITPOINTS] <= 0)
    {
        return;
    };
    var C_NPC her;  her = Hlp_GetNpc(PC_Hero); 
    if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(her))
    {
        return;
    };
    if (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(hero))
    {
        return;
    };
    if (self.guild >= GIL_SEPERATOR_HUM)
    {
        return;
    };   
    
    AI_UnequipArmor     (self);
    AI_EquipBestArmor   (self);
    B_RefreshAtInsert_old();
};
```

**Reply by Kor-Angar** [reply]
в B_RefreshAtInsert можно добавить вызов какой то фунцию которая будет делать "чекин" доспехов и уже в неё засунуть переодевание Грега, Каравалона и прочие переодевалки в зависимости от каких то условий.

**Reply by Kor-Angar** [reply]
![20220922182015_1](https://user-images.githubusercontent.com/49514764/191787073-5174a686-3288-4df0-a20f-9b96e83af54c.jpg)

**Comment by Kor-Angar** [comment]
@dosinabox  а если такую проверку `if (NPC_Pal_lvl(slf, 40) == false)   { return; };`  сделать корректно будет работать?

```D
func int NPC_Pal_lvl(var c_npc slf, var int Pal_lvl)
{   
    if (!Hlp_IsValidNpc(slf))
    { 
        return false;  //npc не существует
    };
    if  slf.attribute[ATR_MANA_MAX] == 199
    {
        return false;   
    };  
    
    ...

    return true;        
};

//*********************************************************//
//
//                      паладины 40 уровня
//
//********************************************************//

func void NPC_Pal_40_lvl_1h(var c_npc slf)
{   
    //----------------
    if (NPC_Pal_lvl(slf, 40) == false)   { return; };
    // ------ Kampf-Talente ------

    ...

};
```

[Npc_Pal_2022.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/10046409/Npc_Pal_2022.zip)

**Reply by dosinabox** [reply]
работать где и с чем?

**Reply by Kor-Angar** [reply]
вот если я вот тут делаю её вызов

```D
func void NPC_Pal_40_lvl_1h(var c_npc slf)
{   
    //----------------
    if (NPC_Pal_lvl(slf, 40) == false)   { return; };
    // ------ Kampf-Talente ------

    ...

};
```

будет ли корректно работать проверка? если например НПС нет.

**Reply by dosinabox** [reply]
а где используется NPC_Pal_40_lvl_1h?

**Reply by Kor-Angar** [reply]
вызывается в 

`B_InitNpcGlobals`  -->  `B_Khorinis_NPC_All`  --> `B_Khorinis_NPC_Pal `  -->  `NPC_Pal_40_lvl_1h`

**Reply by Kor-Angar** [reply]
`B_InitNpcGlobals`  -->  `B_Khorinis_NPC_All`  --> `B_Khorinis_NPC_Pal `  -->  `NPC_Pal_40_lvl_1h`   -->  `NPC_Pal_lvl`
`B_InitNpcGlobals`  -->  `B_Khorinis_NPC_All`  --> `B_Khorinis_NPC_Pal `  -->  `NPC_Pal_40_lvl_2h`   -->  `NPC_Pal_lvl`
`B_InitNpcGlobals`  -->  `B_Khorinis_NPC_All`  --> `B_Khorinis_NPC_Pal `  -->  `NPC_Pal_50_lvl_1h`   -->  `NPC_Pal_lvl`
`B_InitNpcGlobals`  -->  `B_Khorinis_NPC_All`  --> `B_Khorinis_NPC_Pal `  -->  `NPC_Pal_50_lvl_2h`   -->  `NPC_Pal_lvl`

**Reply by Kor-Angar** [reply]
соответственно я `NPC_Pal_lvl` решил сделать функций `int` и добавить проверку на то есть ли НПС в мире.

**Reply by dosinabox** [reply]
каких-то проблем не вижу

**Reply by Kor-Angar** [reply]
я опять же такое решил сделать что бы "снизить" нагрузка при загрузке мира / сейва. или в этом нет практического смысла? то есть являются ли такие "оптимизации" полезными?

**Reply by dosinabox** [reply]
Без тестов и точных цифр об этом нельзя сказать с уверенностью.
А на этапе вызова B_InitNpcGlobals переменная CurrentLevel уже изменена на текущий зен?

**Reply by dosinabox** [reply]
> А на этапе вызова B_InitNpcGlobals переменная CurrentLevel уже изменена на текущий зен?

неа
STARTUP_GLOBAL -> STARTUP_AddonWorld -> INIT_GLOBAL -> INIT_AddonWorld

**Reply by dosinabox** [reply]
Мне всё хочется разобрать B_InitNpcGlobals на отдельные миры:
```cpp
func void B_InitNpcGlobals()
{
	if(CurrentLevel == ADDONWORLD_ZEN)
	{
		Raven = Hlp_GetNpc(BDT_1090_Addon_Raven);
		Logan = Hlp_GetNpc(BDT_1072_Addon_Logan);
		...
	}
	else if(CurrentLevel == NEWWORLD_ZEN)
	{
		Lobart = Hlp_GetNpc(BAU_950_Lobart);
		LobartsBauer1 = Hlp_GetNpc(BAU_955_Bauer);
		LobartsBauer2 = Hlp_GetNpc(BAU_953_Bauer);
		...
	}
	...
```
Если сработает, то будет толк.

**Reply by Kor-Angar** [reply]
> А на этапе вызова B_InitNpcGlobals переменная CurrentLevel уже изменена на текущий зен?
> STARTUP_GLOBAL -> STARTUP_AddonWorld -> INIT_GLOBAL -> INIT_AddonWorld

в MDK да, у тебя кажется нет.

| ------------- | CurrentLevel   | B_InitNpcGlobals |
| ------------- | ------------- | ------------- |
| MDK  | в `CurrentLevel` устанавливается в `STARTUP_МИРА`  и B_ENTER_МИРА |   `B_InitNpcGlobals` вызывается в INIT_МИРА  |
| скиптпачт  | в `CurrentLevel` устанавливается в INIT_МИРА  |  INIT_Global

**Reply by Kor-Angar** [reply]
> Мне всё хочется разобрать B_InitNpcGlobals на отдельные миры:

это же ты сделала. я и НПС проверял в прошлом году. надо внедрять. верни установку CurrentLevel  в STARTUP_МИРА да и всё :-)

**Reply by Kor-Angar** [reply]
а мне казалось что у тебя уже был где то список уже готовый 
https://github.com/dosinabox/g2nr_unofficial_update/discussions/449

**Reply by Kor-Angar** [reply]
ой а в MDK еще и дублируется в **B_ENTER_МИР**.  ну да. Стартап мира же запускается один раз. в стартапе в принципе всё это не нужно. вредно. тогда вызов `B_InitNpcGlobals`  было бы логично делать в `INIT_МИРА`.

**Comment by Kor-Angar** [comment]
@dosinabox  вообщем я разобрался как работает `AI_EquipArmor`. помогли как обычно скрипты Готики 1 MDK :-)

` AI_EquipArmor(hero, item);` - нпс оденет **указанную** броню (даже если она будет не самая лучше). [youtube](https://www.youtube.com/watch?v=NZQ_DxBJNpY)

после завтра попробую переписать свои функции с учётом этого и посмотрю как оно на практике работает.

**Reply by Kor-Angar** [reply]
похоже ей нужна именно глобальная item

так работает.
```D
func void  B_EquipArmor(var c_npc slf,var int Armor)
{
    Npc_GetInvItem  (slf,Armor);
    AI_EquipArmor   (slf, item);
};

```
```
...D
        B_EquipArmor(hero, SLD_ARMOR_M);
        B_EquipArmor(Sld_700_Lee,SLD_ARMOR_L);
...
```

**Reply by Kor-Angar** [reply]
![20221125053329_1](https://user-images.githubusercontent.com/49514764/203888986-c2e1bf85-f97c-4db2-83e3-654099f34b0e.jpg)

**Reply by dosinabox** [reply]
А вне зоны видимости работает?

**Reply by Kor-Angar** [reply]
не тестил ещё

**Reply by Kor-Angar** [reply]
> А вне зоны видимости работает?

нет.

**Comment by Kor-Angar** [comment]
топик про доспехи ОнД https://github.com/dosinabox/g2nr_unofficial_update/discussions/486

**Comment by Kor-Angar** [comment]
счастье у меня привалило. наконец то смогу вынести татушки, текстурки рудокопов и может быть зубы в отдельный проект :-)

**Reply by Kor-Angar** [reply]
Old texture for tattoo and prisoner
 https://steamcommunity.com/sharedfiles/filedetails/?id=2929509312

**Comment by Kor-Angar** [comment]
защита от дураков (меня) :-) 

```D
    func void :B_RefreshAtInsert()
    {
        var string msg;
        
        ...

                if (Npc_HasEquippedArmor(self) == false)
                {
                    Npc_ClearAIQueue(self);
                    AI_EquipBestArmor(self);
                    msg = Str_Format("  --> Npc_HasEquippedArmor --> AI_EquipBestArmor(%s %i)", self.name, self.id);
                    Hlp_PrintConsole(msg);
                };
```
